### PR TITLE
[CALCITE] Iteration statement base and While statement

### PR DESCRIPTION
**Pull request by ArthurInTheShell**
Setup the base iteration statement for future PRs. Implement While Statement.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/SqlIterationStmt.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+public class SqlIterationStmt extends SqlCall {
+  public static final SqlSpecialOperator OPERATOR =
+      new SqlSpecialOperator("ITERATION", SqlKind.ITERATION_STATEMENT);
+
+  public final SqlNode condition;
+  public final SqlStatementList statements;
+  public final SqlIdentifier label;
+
+  /**
+   * Creates a {@code SqlIterationStmt}.
```
Comment (AndrewPochapsky) [inline core/src/main/java/org/apache/calcite/sql/SqlIterationStmt.java:40]: nit: Add new line before params.

File: core/src/main/java/org/apache/calcite/sql/SqlWhileStmt.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * Parse tree for a {@code SqlWhileStmt}.
+ */
+public class SqlWhileStmt extends SqlIterationStmt {
+  public static final SqlSpecialOperator OPERATOR =
+      new SqlSpecialOperator("WHILE", SqlKind.WHILE_STATEMENT);
+
+  /**
+   * Creates a {@code SqlWhileStmt}.
```
Comment (AndrewPochapsky) [inline core/src/main/java/org/apache/calcite/sql/SqlWhileStmt.java:29]: nit: Add new line before params.

File: core/src/main/java/org/apache/calcite/sql/SqlIterationStmt.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      throw SqlUtil.newContextException(beginLabel.getParserPosition(),
+          RESOURCE.beginEndLabelMismatch());
+    }
+    this.condition = Objects.requireNonNull(condition);
+    this.statements = Objects.requireNonNull(statements);
+    this.label = beginLabel;
+  }
+
+  @Override public SqlOperator getOperator() {
+    return null;
```
Comment (AndrewPochapsky) [inline core/src/main/java/org/apache/calcite/sql/SqlIterationStmt.java:66]: Should this return OPERATOR?

File: parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl [hunk 4]
```diff
@@ -4641,6 +4641,11 @@ SqlNode CreateProcedureStmt() :
         e = ConditionalStmt()
     |
         e = CursorStmt()
+    |
+        // This lookahead ensures parser choose
```
Comment (AndrewPochapsky) [inline parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl:4645]: nit: choose -> chooses

File: parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl [hunk 5]
```diff
@@ -5063,3 +5068,44 @@ SqlCloseCursor SqlCloseCursor() :
     <CLOSE> cursorName = SimpleIdentifier()
     { return new SqlCloseCursor(s.end(this), cursorName); }
 }
+
+SqlIterationStmt IterationStmt() :
+{
+    final SqlIterationStmt e;
+}
+{
+    (
```
Comment (AndrewPochapsky) [inline parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl:5077]: nit: Can we get rid of these parentheses?

Comment (ArthurInTheShell) [reply parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl:5077]: These parentheses are setup this way for future PRs to add entries in this statement. Having them here would make the future conflicts of entries easier to resolve.

File: core/src/main/java/org/apache/calcite/sql/SqlIterationStmt.java (hunk 1 above)
Comment (ArthurInTheShell) [reply core/src/main/java/org/apache/calcite/sql/SqlIterationStmt.java:40]: Fixed in https://github.com/googleinterns/calcite/pull/280/commits/f952df71de6a6e78a60de66d49c01ae89b92040e

File: parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl (hunk 5 above)
Comment (AndrewPochapsky) [reply parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl:5077]: Ah ok, SGTM.

File: core/src/main/java/org/apache/calcite/sql/SqlWhileStmt.java (hunk 2 above)
Comment (ArthurInTheShell) [reply core/src/main/java/org/apache/calcite/sql/SqlWhileStmt.java:29]: Fixed in https://github.com/googleinterns/calcite/pull/280/commits/f952df71de6a6e78a60de66d49c01ae89b92040e

File: core/src/main/java/org/apache/calcite/sql/SqlIterationStmt.java (hunk 3 above)
Comment (ArthurInTheShell) [reply core/src/main/java/org/apache/calcite/sql/SqlIterationStmt.java]: Fixed in https://github.com/googleinterns/calcite/pull/280/commits/f952df71de6a6e78a60de66d49c01ae89b92040e

File: parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl (hunk 4 above)
Comment (ArthurInTheShell) [reply parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl]: Good catch. Fixed in https://github.com/googleinterns/calcite/pull/280/commits/f952df71de6a6e78a60de66d49c01ae89b92040e

File: core/src/main/java/org/apache/calcite/sql/SqlIterationStmt.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+import java.util.List;
+import java.util.Objects;
+
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/**
+ * Parse tree for a {@code SqlIterationStmt}.
+ */
+public class SqlIterationStmt extends SqlCall {
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/SqlIterationStmt.java:31]: Should this be an abstract class? Should anyone be able to instantiate a SqlIterationStmt directly?

File: parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl [hunk 7]
```diff
@@ -4641,6 +4641,11 @@ SqlNode CreateProcedureStmt() :
         e = ConditionalStmt()
     |
         e = CursorStmt()
+    |
+        // This lookahead ensures parser chooses
+        // the right path when facing begin label.
```
Comment (dasch-1) [inline parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl:4646]: nit: Seems like this line was wrapped to soon and more can fit on the previous line.

File: core/src/main/java/org/apache/calcite/sql/SqlIterationStmt.java (hunk 6 above)
Comment (ArthurInTheShell) [reply core/src/main/java/org/apache/calcite/sql/SqlIterationStmt.java]: Yes, this should be an abstract class instead. Fixed in https://github.com/googleinterns/calcite/pull/280/commits/b9f8fa4c7b3b2c5379d302ff26031064b0ed9af9

File: parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl (hunk 7 above)
Comment (ArthurInTheShell) [reply parsing/dialects/intermediate/dialects/dialect1/parserImpls.ftl]: line wrap fixed in https://github.com/googleinterns/calcite/pull/280/commits/b9f8fa4c7b3b2c5379d302ff26031064b0ed9af9

**Pull request merged**